flink-state-backends/flink-statebackend-heap-spillable/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMap.java

```diff
@@ -0,0 +1,1505 @@
+			DEFAULT_MAX_KEYS_TO_DELETE_ONE_TIME, DEFAULT_LOGICAL_REMOVED_KEYS_RATIO);
+	}
+
+	public CopyOnWriteSkipListStateMap(
+			TypeSerializer<K> keySerializer,
+			TypeSerializer<N> namespaceSerializer,
+			TypeSerializer<S> stateSerializer,
+			Allocator spaceAllocator,
+			int numKeysToDeleteOneTime,
+			float logicalRemovedKeysRatio) {
+		this.skipListKeySerializer = new SkipListKeySerializer<>(keySerializer, namespaceSerializer);
+		this.skipListValueSerializer = new SkipListValueSerializer<>(stateSerializer);
+		this.spaceAllocator = Preconditions.checkNotNull(spaceAllocator);
+		Preconditions.checkArgument(numKeysToDeleteOneTime >= 0,
+			"numKeysToDeleteOneTime should be non-negative, but is "  + numKeysToDeleteOneTime);
+		this.numKeysToDeleteOneTime = numKeysToDeleteOneTime;
+		Preconditions.checkArgument(logicalRemovedKeysRatio >= 0 && logicalRemovedKeysRatio <= 1,
+			"logicalRemovedKeysRatio should be in [0, 1], but is " + logicalRemovedKeysRatio);
+		this.logicalRemovedKeysRatio = logicalRemovedKeysRatio;
+
+		this.levelIndexHeader = new OnHeapLevelIndexHeader();
+		this.randomSeed = ThreadLocalRandom.current().nextInt() | 0x0100;
+
+		this.stateMapVersion = 0;
+		this.highestRequiredSnapshotVersion = 0;
+		this.highestFinishedSnapshotVersion = 0;
+		this.snapshotVersions = new TreeSet<>();
+
+		this.totalSize = 0;
+		this.requestCount = 0;
+
+		this.logicallyRemovedNodes = new HashSet<>();
+		this.pruningValueNodes = ConcurrentHashMap.newKeySet();
+
+		this.closed = new AtomicBoolean(false);
+		this.resourceGuard = new ResourceGuard();
+	}
+
+	@Override
+	public int size() {
+		return totalSize - logicallyRemovedNodes.size();
+	}
+
+	/**
+	 * Returns total size of this map, including logically removed state.
+	 */
+	public int totalSize() {
+		return totalSize;
+	}
+
+	public int getRequestCount() {
+		return requestCount;
+	}
+
+	@Override
+	public S get(K key, N namespace) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+
+		long node = getNode(keyByteBuffer, 0, keyLen);
+
+		if (node == NIL_NODE) {
+			return null;
+		}
+
+		return helpGetNodeState(node);
+	}
+
+	@Override
+	public boolean containsKey(K key, N namespace) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+
+		long node = getNode(keyByteBuffer, 0, keyLen);
+
+		return node != NIL_NODE;
+	}
+
+	@Override
+	public void put(K key, N namespace, S state) {
+		updateStat();
+		byte[] keyBytes = skipListKeySerializer.serialize(key, namespace);
+		ByteBuffer keyByteBuffer = ByteBuffer.wrap(keyBytes);
+		int keyLen = keyBytes.length;
+		byte[] value = skipListValueSerializer.serialize(state);
+
+		putNode(keyByteBuffer, 0, keyLen, value, false);
```

Comment:
Instead of passing `keyLen` into `putNode`, `getNode` and `removeNode`, couldn't we also calculate this value from `keyByteBuffer` inside of these methods? That way we would not have to calculate it always before calling one of these methods.

Reply:
Ok, I see why we are calculating the `keyLen` because the passed in `ByteBuffer` must not necessarily be fully used. Maybe `getNode` does not need to accept `ByteBuffers` but can directly work with `byte[]`.

Reply:
Currently `getNode` depends on the util methods of `SkipListUtils`, which takes `ByteBuffer` as the input parameter, so changing to directly pass `byte[]` in may not adapt. Will check the cost of `ByteBuffer.wrap` through jmh and get back later.